mcs/class/System.Windows.Forms/System.Windows.Forms.RTF/RTF.cs

```diff
@@ -369,49 +363,52 @@ private char GetChar(bool skipCrLf)
 			GetToken2();
 
 			if (this.rtf_class == TokenClass.Text) {
-				this.minor = (Minor)this.cur_charset[(int)this.major];
-				if (encoding == null) {
+				if (encoding == null)
 					encoding = Encoding.GetEncoding (encoding_code_page);
-				}
 				encoded_text = new String (encoding.GetChars (new byte [] { (byte) this.major }));
-			}
-
-			if (this.cur_charset.Flags == CharsetFlags.None) {
-				return this.rtf_class;
-			}
-
-			if (CheckCMM (TokenClass.Control, Major.Unicode, Minor.UnicodeAnsiCodepage)) {
+			} else if (CheckCMM (TokenClass.Control, Major.Unicode, Minor.UnicodeAnsiCodepage)) {
 				encoding_code_page = param;
 
 				// fallback to the default one in case we have an invalid value
 				if (encoding_code_page < 0 || encoding_code_page > 65535)
 					encoding_code_page = DefaultEncodingCodePage;
-			}
 
-			if (((this.cur_charset.Flags & CharsetFlags.Read) != 0) && CheckCM(TokenClass.Control, Major.CharSet)) {
-				this.cur_charset.ReadMap();
-			} else if (((this.cur_charset.Flags & CharsetFlags.Switch) != 0) && CheckCMM(TokenClass.Control, Major.CharAttr, Minor.FontNum)) {
+				encoding = null;
+			} else if (CheckCMM(TokenClass.Control, Major.CharAttr, Minor.FontNum)) {
 				Font	fp;
 
-				fp = Font.GetFont(this.font_list, this.param);
+				fp = Font.GetFont(this.fonts, this.param);
 
 				if (fp != null) {
-					if (fp.Name.StartsWith("Symbol")) {
-						this.cur_charset.ID = CharsetType.Symbol;
+					if (fp.Codepage != 0) {
+						if (fp.Codepage != encoding_code_page) {
+							encoding_code_page = fp.Codepage;
+							encoding = null;
+						}
 					} else {
-						this.cur_charset.ID = CharsetType.General;
-					}
-				} else if (((this.cur_charset.Flags & CharsetFlags.Switch) != 0) && (this.rtf_class == TokenClass.Group)) {
-					switch(this.major) {
-						case Major.BeginGroup: {
-							this.charset_stack.Push(this.cur_charset);
-							break;
+						var cp = CharsetToCodepage.Translate (fp.Charset);
+						if (cp != 0 && cp != encoding_code_page) {
+							encoding_code_page = cp;
+							encoding = null;
 						}
+					}
+				}
+			} else if (this.rtf_class == TokenClass.Group) {
+				switch(this.major) {
+					case Major.BeginGroup: {
+						charset_stack.Push(encoding);
```

Comment:
This looks wrong since `encoding` is computed lazily so it can push `null` to the stack which results in incorrect state after `EndGroup`. The stack should either keep the whole `encoding`+`encoding_code_page` pair or only `encoding_code_page`.

Reply:
Good point. I've changed it to storing `encoding_code_page` as you suggest (keeping the pair would be nicer, but is more bother and would probably be harder to read).

On a side note, should I push these changes immediately, or wait a day or two in case there are more so I can push them all at once?

Reply:
> On a side note, should I push these changes immediately, or wait a day or two in case there are more so I can push them all at once?

I'd say push them now. I went through the rest of changes and didn't find anything strikingly wrong but I will probably do one more review pass later in the week.

Reply:
OK, done. I hope this looks better.